### ISSUE TITLE
NumPy throws out-of-memory error when running training code for model 1

### ISSUE THREAD

**Issue by generic-github-user** [usnistgov/STVM_NLP_Research#3]
> NumPy is throwing an out of memory error due to trying to create an array with each sample on one axis and each word in the vocabulary list on the other (~16GB in total). I am thinking that the way I downloaded/prepared the dataset may have caused this, but the paper mentions the models having been originally trained on a GPU cluster with 32GB of RAM per unit, so this might be intended. If that is the case, perhaps it is worth including a batch size option for users trying to run the training code for themselves so that the model(s) can be trained incrementally on smaller samples of the original dataset?

For reference, I have 8GB of RAM (and am training on CPU due to CUDA installation issues I am still trying to resolve). If adding an option for training on machines with less compute/memory would be beneficial, I could implement a (backwards-compatible) function that either loads smaller batches and trains on them separately, downscales the dataset size, or stores intermediate results on disk instead of in memory.

I believe the array used for the training data is a one-hot encoding over the entire IMDB dataset vocabulary (and is thus almost entirely zeros), so a sparse matrix/tensor that stores only the indices might be more efficient. I believe TensorFlow also has some dataset classes and pipelines for processing/training on large datasets, though I am not well-versed in these features.

**Comment by mhasan08** [comment]
I agree with you that loading such large array at once might not be good way. One way is to look into tf data pipeline, there are other ways too.
